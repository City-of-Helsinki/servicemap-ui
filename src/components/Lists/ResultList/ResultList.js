/* eslint-disable camelcase */
import React from 'react';
import PropTypes from 'prop-types';
import {
  List, Typography, Divider,
} from '@material-ui/core';
import { FormattedMessage } from 'react-intl';
import UnitItem from '../../ListItems/UnitItem';
import ServiceItem from '../../ListItems/ServiceItem';
import AddressItem from '../../ListItems/AddressItem';

class ResultList extends React.Component {
  // Update only when data changes
  shouldComponentUpdate(nextProps) {
    const { data, title } = this.props;
    return (data !== nextProps.data) || (title !== nextProps.title);
  }

  render() {
    const {
      classes, data, listId, resultCount, title, titleComponent,
    } = this.props;

    return (
      <div className={classes.root}>
        {
          title
          && (
            <div className={classes.title}>
              <div className={classes.titleContainer}>
                <Typography
                  id={`${listId}-result-title`}
                  className={`${classes.left} SearchResultTitle`}
                  component={titleComponent}
                  variant="subtitle1"
                  aria-labelledby={`${listId}-result-title ${listId}-result-title-info`}
                  tabIndex="-1"
                >
                  {title}

                </Typography>
                <Typography
                  id={`${listId}-result-title-info`}
                  className={classes.right}
                  component="p"
                  variant="caption"
                  aria-hidden="true"
                >
                  <FormattedMessage id="search.resultList" values={{ count: resultCount || data.length }} />
                </Typography>
              </div>
            </div>
          )
        }
        <Divider aria-hidden="true" />
        <List className={classes.list} id={listId}>
          {
            data && data.length
            && data.map((item) => {
              const { id, object_type, sort_index } = item;
              // Figure out correct icon for item
              let itemComponent = null;
              switch (object_type) {
                case 'unit':
                  itemComponent = <UnitItem key={`unit-${id}`} unit={item} listId={listId} single />;
                  break;
                case 'service':
                  itemComponent = <ServiceItem key={`service-${id}`} service={item} />;
                  break;
                case 'address':
                  itemComponent = <AddressItem key={`address-${sort_index}`} address={item} />;
                  break;
                default:
                  itemComponent = null;
              }
              if (item) {
                return itemComponent;
              }
              return null;
            })
          }
        </List>
      </div>
    );
  }
}

export default ResultList;

// Typechecking
ResultList.propTypes = {
  classes: PropTypes.objectOf(PropTypes.any),
  data: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)).isRequired,
  listId: PropTypes.string.isRequired,
  resultCount: PropTypes.number,
  title: PropTypes.string,
  titleComponent: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
};

ResultList.defaultProps = {
  classes: {},
  resultCount: null,
  title: null,
  titleComponent: 'h3',
};
