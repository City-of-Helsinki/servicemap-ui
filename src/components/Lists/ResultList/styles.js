export default theme => ({
  root: {
    height: 'auto',
    overflowY: 'auto',
    flex: '1 1 auto',
    maxWidth: '100%',
    overflowX: 'hidden',
  },
  title: {
    width: '100%',
    display: 'flex',
  },
  titleContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
    overflow: 'hidden',
    padding: theme.spacing(1, 2),
  },
  titleText: {
    ...theme.typography.body2,
    textTransform: 'uppercase',
    fontWeight: 'bold',
  },
  left: {
    float: 'left',
    margin: theme.spacing(1),
  },
  right: {
    float: 'right',
    margin: theme.spacing(1),
  },
  list: {
    maxHeight: '100%',
  },
});
