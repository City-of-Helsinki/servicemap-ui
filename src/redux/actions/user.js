import LocalStorageUtility from '../../utils/localStorage';
import fetchAddress from '../../views/MapView/utils/fetchAddress';

export const setLocale = locale => ({
  type: 'SET_LOCALE',
  locale,
});

export const setInitialLoad = () => ({
  type: 'SET_INITIAL_LOAD',
});

export const setPage = page => ({
  type: 'SET_CURRENT_PAGE',
  page,
});

export const setUserPosition = position => ({
  type: 'SET_POSITION',
  position,
});

const setTheme = theme => ({
  type: 'SET_THEME',
  theme,
});


export const changeLocaleAction = locale => async (dispatch) => {
  dispatch(setLocale(locale));
};

export const actionSetInitialLoad = () => async (dispatch) => {
  dispatch(setInitialLoad());
};

export const setCurrentPage = page => async (dispatch) => {
  dispatch(setPage(page));
};

export const changeTheme = theme => async (dispatch) => {
  LocalStorageUtility.saveItem('theme', theme);
  dispatch(setTheme(theme));
};

export const findUserLocation = () => async (dispatch) => {
  const success = (position) => {
    if (position.coords.accuracy < 10000) {
      fetchAddress({ lat: position.coords.latitude, lng: position.coords.longitude })
        .then((data) => {
          dispatch(setUserPosition({
            coordinates: position.coords,
            allowed: true,
            addressData: data,
          }));
        });
    } else {
      console.warn(`Position accuracy: ${position.coords.accuracy}. Max accuracy: 10000`);
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  const error = (err) => {
    console.warn(`GeoLocation error:(${err.code}): ${err.message}`);
    if (err.code === 1) {
      dispatch(setUserPosition({ coordinates: null, allowed: false, addressData: null }));
    } else {
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  navigator.geolocation.getCurrentPosition(success, error);
};


export default { setLocale, changeLocaleAction };
