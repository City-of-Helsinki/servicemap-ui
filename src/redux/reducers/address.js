const initialState = {
  units: [],
  addressCoordinates: null,
  addressData: null,
  adminDistricts: [],
  toRender: 'adminDistricts',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_ADDRESS_DATA':
      return {
        ...state,
        addressData: action.data,
      };
    case 'SET_ADDRESS_LOCATION':
      return {
        ...state,
        addressCoordinates: action.location.addressCoordinates,
      };
    case 'SET_ADDRESS_UNITS':
      return {
        ...state,
        units: action.units,
      };
    case 'SET_ADMINISTRATIVE_DISTRICTS':
      return {
        ...state,
        adminDistricts: action.data,
      };
    case 'SET_TO_RENDER':
      return {
        ...state,
        toRender: action.data,
      };
    default:
      return state;
  }
};
