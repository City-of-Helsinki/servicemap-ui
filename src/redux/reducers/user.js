
const initialState = {
  initialLoad: false,
  locale: 'fi',
  page: 'home',
  theme: 'default',
  position: {
    coordinates: null,
    allowed: false,
  },
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_LOCALE':
      return {
        ...state,
        locale: action.locale,
      };
    case 'SET_INITIAL_LOAD':
      return {
        ...state,
        initialLoad: true,
      };
    case 'SET_CURRENT_PAGE':
      return {
        ...state,
        page: action.page,
      };
    case 'SET_POSITION':
      return {
        ...state,
        position: action.position,
      };
    case 'SET_THEME':
      return {
        ...state,
        theme: action.theme,
      };
    default:
      return state;
  }
};
