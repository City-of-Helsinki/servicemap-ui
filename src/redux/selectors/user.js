
export const getTheme = state => state.user.theme;
export const getPage = state => state.user.page;
export const getLocale = state => state.user.locale;
