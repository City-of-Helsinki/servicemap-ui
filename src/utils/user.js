import { useSelector } from 'react-redux'
 
export const useUserTheme = () => useSelector(state => state.user.theme);
export const useUserLocale = () => useSelector(state => state.user.locale);
