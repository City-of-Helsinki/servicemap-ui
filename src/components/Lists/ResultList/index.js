import ResultList from './ResultList';

export default ResultList;
