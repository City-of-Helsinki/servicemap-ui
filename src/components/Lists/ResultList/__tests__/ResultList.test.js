// Link.react.test.js
import React from 'react';
import ResultList from '../ResultList';
import { initialState } from '../../../../redux/reducers/user';
import { getRenderWithProviders } from '../../../../../jestUtils';

const mockData = [
  {
    accessibility_properties: [],
    accessibility_shortcoming_count: {},
    contract_type: {
      id: 'municipal_service',
      description: { fi: 'kunnallinen palvelu', sv: 'kommunal tjänst', en: 'municipal service' },
    },
    id: 63115,
    municipality: 'espoo',
    name: { fi: 'Lippulaivan kirjasto', sv: 'Lippulaivabiblioteket', en: 'Lippulaiva library' },
    object_type: 'unit',
    street_address: { fi: 'Merikarhunkuja 11', sv: 'Sjöbjörnsgränden 11', en: 'Merikarhunkuja 11' },
  },
  {
    accessibility_properties: [],
    accessibility_shortcoming_count: {},
    contract_type: {
      id: 'municipal_service',
      description: { fi: 'kunnallinen palvelu', sv: 'kommunal tjänst', en: 'municipal service' },
    },
    id: 62032,
    municipality: 'vantaa',
    name: {
      fi: 'Kivistön kirjasto',
      sv: 'Kivistö bibliotek',
      en: 'Kivistö Library',
    },
    object_type: 'unit',
    street_address: { fi: 'Topaasikuja 11', sv: 'Topasgränden 11', en: 'Topaasikuja 11' },
  },
];

// Generic required props for SimpleListItem
const mockProps = {
  beforeList: <p>Test before list</p>,
  customComponent: null,
  data: mockData,
  listId: 'testID',
  resultCount: 5,
  title: 'Title text',
  titleComponent: 'h3',
};

const renderWithProviders = getRenderWithProviders({
  user: initialState,
  settings: {},
});

describe('<ResultList />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<ResultList {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does render list', () => {
    const { getAllByRole } = renderWithProviders(<ResultList {...mockProps} />);
    const count = getAllByRole('link', { selector: 'li' }).length;
    expect(count === 2).toBeTruthy();
  });

  it('does render beforeList', () => {
    const { getByText } = renderWithProviders(<ResultList {...mockProps} />);
    const text = getByText('Test before list').textContent;
    expect(text).toEqual('Test before list');
  });

  it('does render accessibility attributes correctly', () => {
    const { getAllByRole } = renderWithProviders(<ResultList {...mockProps} />);
    const items = getAllByRole('link', { selector: 'li' });
    const firstItem = items[0];
    const firstItemSRText = firstItem.querySelectorAll('p')[0];
    const firstItemResultTitle = firstItem.querySelectorAll('p')[1];

    // List item's image should be aria-hidden
    expect(firstItem.querySelector('img').getAttribute('aria-hidden')).toBeTruthy();
    expect(firstItem.getAttribute('role')).toEqual('link');
    expect(firstItem.getAttribute('tabIndex')).toEqual('0');
    expect(firstItemSRText.className.indexOf('ResultItem-title') > 0).toBeTruthy();
    expect(firstItemResultTitle.className.indexOf('ResultItem-title') > 0).toBeTruthy();
    expect(firstItemResultTitle.getAttribute('aria-hidden')).toBeTruthy();
  });
});
