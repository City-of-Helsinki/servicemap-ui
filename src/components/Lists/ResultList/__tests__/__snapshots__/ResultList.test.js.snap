// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultList /> should work 1`] = `
<div>
  <div>
    <div>
      <div>
        <h3
          aria-labelledby="testID-result-title testID-result-title-info"
          class="MuiTypography-root MuiTypography-subtitle1 undefined undefined SearchResultTitle css-dlbwt1-MuiTypography-root"
          id="testID-result-title"
          tabindex="-1"
        >
          Title text
        </h3>
        <p
          aria-hidden="true"
          class="MuiTypography-root MuiTypography-caption css-r3hlsg-MuiTypography-root"
          id="testID-result-title-info"
        >
          5 matches
        </p>
      </div>
    </div>
    <hr
      aria-hidden="true"
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <p>
      Test before list
    </p>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      id="testID"
    >
      <li
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button ResultItem-listItem-3 css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
        role="link"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root ResultItem-listItemIcon-4 css-cveggr-MuiListItemIcon-root"
        >
          <img
            alt=""
            aria-hidden="true"
            class="Connect(UnitIcon)-icon-21 "
            src="data:image/png;base64,00"
          />
        </div>
        <div
          class="ResultItem-itemTextContainer-10  "
        >
          <div
            class="ResultItem-topRow-12"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-8 ResultItem-srOnly css-1ak20dk-MuiTypography-root"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
            >
              
    Lippulaivan kirjasto 
    Kunnallinen palvelu 
     
     
    
  
            </p>
            <p
              aria-hidden="true"
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-8  Connect(UnitItem)-title-1  ResultItem-title css-1ak20dk-MuiTypography-root"
              role="textbox"
            >
              Lippulaivan kirjasto
            </p>
          </div>
          <div>
            <div>
              <p
                aria-hidden="true"
                class="MuiTypography-root MuiTypography-caption ResultItem-noMargin-9 ResultItem-text-17  ResultItem-subtitle css-r3hlsg-MuiTypography-root"
              >
                Kunnallinen palvelu
              </p>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
      >
        <hr
          class="MuiDivider-root MuiDivider-inset ResultItem-divider-18 css-uls1yk-MuiDivider-root"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button ResultItem-listItem-3 css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
        role="link"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root ResultItem-listItemIcon-4 css-cveggr-MuiListItemIcon-root"
        >
          <img
            alt=""
            aria-hidden="true"
            class="Connect(UnitIcon)-icon-21 "
            src="data:image/png;base64,00"
          />
        </div>
        <div
          class="ResultItem-itemTextContainer-10  "
        >
          <div
            class="ResultItem-topRow-12"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-8 ResultItem-srOnly css-1ak20dk-MuiTypography-root"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
            >
              
    Kivistön kirjasto 
    Kunnallinen palvelu 
     
     
    
  
            </p>
            <p
              aria-hidden="true"
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-8  Connect(UnitItem)-title-1  ResultItem-title css-1ak20dk-MuiTypography-root"
              role="textbox"
            >
              Kivistön kirjasto
            </p>
          </div>
          <div>
            <div>
              <p
                aria-hidden="true"
                class="MuiTypography-root MuiTypography-caption ResultItem-noMargin-9 ResultItem-text-17  ResultItem-subtitle css-r3hlsg-MuiTypography-root"
              >
                Kunnallinen palvelu
              </p>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
      >
        <hr
          class="MuiDivider-root MuiDivider-inset ResultItem-divider-18 css-uls1yk-MuiDivider-root"
        />
      </li>
    </ul>
  </div>
</div>
`;
